### PR TITLE
Handle Russian and Ukrainian plurals

### LINKED ISSUE #1968: Translations: support number-driven declension
While reviewing PR #1962, I saw the following in the Russian translation:

```
состоит из {{ supported_chapters | length }} глав
```

The word глав (chapter) would take a different ending if the number preceding it ended in 1 (but not 11).

This is unlikely to be the only instance where we need to decline the noun differently depending on the number and this issue would affect Ukrainian, too.  (As well as probably all other Slavic languages were we to have translations for them.)

### PR DESCRIPTION
**Pull request by bazzadp**
Fixes #1968 using [plural-ru](https://github.com/andrewp-as-is/plural-ru.py) function. Note this is not available as an importable function so added code directly to our code base (checked the licence first).

Home page intro:
Home page intro

Methodology text:
Methodology text

Contributors:
Contributors

@dtikhonov can you confirm I've used the right words?

English word|ends in 1 (except 11)| ends in 2,3 and 4 (except 12, 13, 14)|else
-----|-----|----|----
chapters|глава|главы|глав
contributors|участник|участника|участников
people|человек|человека|человек

It does look like we used "главах" below but Google Translate didn't come up with that at all so I guess double check that?

@tymosh does Ukrainian follow the same rules? Should we do the same for those templates?

### CODE REVIEW

**Comment by tymosh** [comment]
Hi @bazzadp. Yes, the same mechanism works for Ukrainian:

English word|ends in 1 (except 11)| ends in 2,3 and 4 (except 12, 13, 14)|else
-----|-----|----|----
chapters|розділ|розділи|розділів
contributors|контриб'ютор|контриб'ютори|контриб'юторів
people|людина|людини|людей

I will also need to check sentences for correct words endings in context.

**Comment by bazzadp** [comment]
Thanks @tymosh . Added those in [the last commit](https://github.com/HTTPArchive/almanac.httparchive.org/pull/1969/commits/f091286bdccdff85d123ca7e5ecd6ae1759b0c30) if you wanna double check.

Also added test site here: https://20210204t112250-dot-webalmanac.uk.r.appspot.com

File: src/templates/uk/base.html [hunk 1]
```diff
@@ -20,7 +20,7 @@
 
 {% block mission %}
 <p>
-Наша місія - поєднати суху статистику і тренди від HTTP Archive з досвідом веб-спільноти. Web Almanac це всеосяжний звіт про стан вебу, підтриманий реальними даними та довіреними веб-експертами. Видання {{ year }} року складається з {{ supported_chapters | length }} розділів, які охоплюють аспекти контенту сторінки, досвіду користування, публікації та розповсюдження.
+Наша місія - поєднати суху статистику і тренди від HTTP Archive з досвідом веб-спільноти. Web Almanac це всеосяжний звіт про стан вебу, підтриманий реальними даними та довіреними веб-експертами. Видання {{ year }} року складається з {{ supported_chapters | length }} {{ plural_ru(supported_chapters | length | int, ["розділ", "розділи", "розділів"]) }}, які охоплюють аспекти контенту сторінки, досвіду користування, публікації та розповсюдження.
```
Comment (tymosh) [inline src/templates/uk/base.html:23]:
```suggestion
Наша місія - поєднати суху статистику і тренди від HTTP Archive з досвідом веб-спільноти. Web Almanac - це всеосяжний звіт про стан вебу, підтриманий реальними даними та довіреними веб-експертами. Видання {{ year }} року складається з {{ supported_chapters | length }} {{ plural_ru(supported_chapters | length | int, ["розділу", "розділів", "розділів"]) }}, які охоплюють аспекти контенту сторінки, досвіду користування, публікації та розповсюдження.
```

File: src/templates/uk/base.html [hunk 2]
```diff
@@ -117,7 +117,7 @@
...
 {% block methodology_description %}
-Якщо не вказано інше, метрики в усіх {{ supported_chapters | length }} розділах Web Almanac {{ year }} взяті з датасету HTTP Archive. HTTP Archive – це громадський проект, що відслідковує, як будується веб з 2010 року. Метадата про {{ self.total_websites() }} вебсайтів тестується щомісяця, використовуючи WebPageTest та Lighthouse, і додається до публічної бази даних BigQuery для аналізу. Для метрик Web Almanac {{ year }} був використаний як базис набір даних від {{ self.dataset() }} року. Більше інформації на сторінці Методологія.
+Якщо не вказано інше, метрики в усіх {{ supported_chapters | length }} {{ plural_ru(supported_chapters | length | int, ["розділ", "розділи", "розділів"]) }} Web Almanac {{ year }} взяті з датасету HTTP Archive. HTTP Archive – це громадський проект, що відслідковує, як будується веб з 2010 року. Метадата про {{ self.total_websites() }} вебсайтів тестується щомісяця, використовуючи WebPageTest та Lighthouse, і додається до публічної бази даних BigQuery для аналізу. Для метрик Web Almanac {{ year }} був використаний як базис набір даних від {{ self.dataset() }} року. Більше інформації на сторінці Методологія.
```
Comment (tymosh) [inline src/templates/uk/base.html:120]:
```suggestion
Якщо не вказано інше, метрики в усіх {{ supported_chapters | length }} {{ plural_ru(supported_chapters | length | int, ["розділі", "розділах", "розділах"]) }} Web Almanac {{ year }} взяті з датасету HTTP Archive. HTTP Archive – це громадський проект, що відслідковує, як будується веб з 2010 року. Метадата про {{ self.total_websites() }} вебсайтів тестується щомісяця, використовуючи WebPageTest та Lighthouse, і додається до публічної бази даних BigQuery для аналізу. Для метрик Web Almanac {{ year }} був використаний як базис набір даних від {{ self.dataset() }} року. Більше інформації на сторінці Методологія.
```

File: src/templates/uk/2019/contributors.html [hunk 3]
```diff
@@ -2,9 +2,9 @@
 
 {% block title %}Контриб'ютори 2019 року | Web Almanac від HTTP Archive{% endblock %}
 
-{% block description %}{{ config.contributors.items() | length }} неймовірних людей, які долучилися до створення Web Almanac 2019 як аналітики, автори, дизайнери, розробники, рецензенти, керівники, редактори та перекладачі.{% endblock %}
+{% block description %}{{ config.contributors.items() | length }} неймовірних {{ plural_ru(config.contributors.items() | length | int, ["людина", "людини", "людей"]) }}, які долучилися до створення Web Almanac 2019 як аналітики, автори, дизайнери, розробники, рецензенти, керівники, редактори та перекладачі.{% endblock %}
```
Comment (tymosh) [inline src/templates/uk/2019/contributors.html:5]:
```suggestion
{% block description %}{{ config.contributors.items() | length }} {{ plural_ru(config.contributors.items() | length | int, ["неймовірна людина", "неймовірні людини", "неймовірних людей"]) }}, які долучилися до створення Web Almanac 2019 як аналітики, автори, дизайнери, розробники, рецензенти, керівники, редактори та перекладачі.{% endblock %}
```

File: src/templates/uk/2020/contributors.html [hunk 4]
```diff
@@ -2,9 +2,9 @@
 
 {% block title %}Контриб'ютори {{ year }} року | Web Almanac від HTTP Archive{% endblock %}
 
-{% block description %}{{ config.contributors.items() | length }} неймовірних людей, які долучилися до створення Web Almanac {{ year }} як аналітики, автори, дизайнери, розробники, рецензенти, керівники, редактори та перекладачі.{% endblock %}
+{% block description %}{{ config.contributors.items() | length }} неймовірних {{ plural_ru(config.contributors.items() | length | int, ["людина", "людини", "людей"]) }},, які долучилися до створення Web Almanac {{ year }} як аналітики, автори, дизайнери, розробники, рецензенти, керівники, редактори та перекладачі.{% endblock %}
```
Comment (tymosh) [inline src/templates/uk/2020/contributors.html:5]:
```suggestion
{% block description %}{{ config.contributors.items() | length }} {{ plural_ru(config.contributors.items() | length | int, ["неймовірна людина", "неймовірні людини", "неймовірних людей"]) }}, які долучилися до створення Web Almanac {{ year }} як аналітики, автори, дизайнери, розробники, рецензенти, керівники, редактори та перекладачі.{% endblock %}
```

**Review by tymosh — CHANGES_REQUESTED** [review]
Here's another thing: there are different word endings for "contributors" in constructions "124 contributors" and "5 of 124 contributors" in Ukrainian. Is there any way to do that?

File: src/templates/ru/2020/contributors.html [hunk 5]
```diff
@@ -2,9 +2,9 @@
...
+{% block description %}{{ config.contributors.items() | length }} {{ plural_ru(config.contributors.items() | length | int, ["человек", "человека", "человек"]) }}, внесших свой вклад в Web Almanac за {{ year }} год — это аналитики, авторы, дизайнеры, разработчики, редакторы, руководители, проверяющие и переводчики.{% endblock %}
 
-{% block filter_by_team %}Фильтр по ролям в команде: <span id="filtered-contributors">{{ self.contributors() }}</span><span id="contributors-total-text" class="hidden"> из <span id="contributors-total">{{ config.contributors.items() | length }}</span></span> участников.{% endblock %}
+{% block filter_by_team %}Фильтр по ролям в команде: <span id="filtered-contributors">{{ self.contributors() }}</span><span id="contributors-total-text" class="hidden"> из <span id="contributors-total">{{ config.contributors.items() | length }}</span></span> {{ plural_ru(config.contributors.items() | length | int, ["участник", "участника", "участников"]) }}.{% endblock %}
```
Comment (dtikhonov) [inline src/templates/ru/2020/contributors.html:7]: You need genitive here:
> "участника", "участников", "участников"

File: src/templates/ru/base.html [hunk 6]
```diff
@@ -20,7 +20,7 @@
 
 {% block mission %}
 <p>
-Наша задача — объединить необработанные статистические данные и тенденции HTTP Archive вместе со знаниями веб-сообщества. Web Almanac — это всеобъемлющий отчёт о состоянии веба, подкреплённый реальными данными и подтверждённый авторитетными веб-экспертами. Издание {{ year }} года состоит из {{ supported_chapters | length }} глав, охватывающих такие аспекты, как содержание страниц, пользовательские ощущения, публикация и распространение информации.
+Наша задача — объединить необработанные статистические данные и тенденции HTTP Archive вместе со знаниями веб-сообщества. Web Almanac — это всеобъемлющий отчёт о состоянии веба, подкреплённый реальными данными и подтверждённый авторитетными веб-экспертами. Издание {{ year }} года состоит из {{ supported_chapters | length }} {{ plural_ru(supported_chapters | length | int, ["глава", "главы", "глав"]) }}, охватывающих такие аспекты, как содержание страниц, пользовательские ощущения, публикация и распространение информации.
```
Comment (dtikhonov) [inline src/templates/ru/base.html:23]: Same here:
> "главы", "глав", "глав"

File: src/templates/ru/base.html [hunk 7]
```diff
@@ -117,7 +117,7 @@
...
+Если не указано иное, метрики во всех {{ supported_chapters | length }} {{ plural_ru(supported_chapters | length | int, ["глава", "главы", "глав"]) }} Web Almanac {{ year }} получены из набора данных HTTP Archive. HTTP Archive — это управляемый сообществом проект по исследованию веба с 2010 года. При помощи WebPageTest и Lighthouse ежемесячно проверяются метаданные {{ self.total_websites() }} сайтов, которые затем вносятся в общедоступную базу данных BigQuery для последующего анализа. В качестве основы для метрик Web Almanac {{ year }} был использован набор данных за {{ self.dataset() }} года. Дополнительную информацию смотрите на странице по методологии.
```
Comment (dtikhonov) [inline src/templates/ru/base.html:120]: Here, the translation does not consider that there may be just one chapter: the word всех (all) cannot be used with 1!  (E.g. "in all one chapters"... does not make sense.)

Nevertheless, I think we can safely ignore the possibility that the Almanac has only one chapter.

The forms of the words you want are
> "глава", "главах", "главах"

**Review by dtikhonov — CHANGES_REQUESTED** [review]

File: src/templates/ru/2019/contributors.html [hunk 8]
```diff
@@ -2,9 +2,9 @@
...
+{% block description %}{{ config.contributors.items() | length }} {{ plural_ru(config.contributors.items() | length | int, ["человек", "человека", "человек"]) }}, внесших свой вклад в Web Almanac за 2019 год — это аналитики, авторы, участники мозговых штурмов, дизайнеры, разработчики, редакторы, руководители, проверяющие и переводчики.{% endblock %}
 
-{% block filter_by_team %}Фильтр по ролям в команде: <span id="filtered-contributors">{{ self.contributors() }}</span><span id="contributors-total-text" class="hidden"> из <span id="contributors-total">{{ config.contributors.items() | length }}</span></span> участников.{% endblock %}
+{% block filter_by_team %}Фильтр по ролям в команде: <span id="filtered-contributors">{{ self.contributors() }}</span><span id="contributors-total-text" class="hidden"> из <span id="contributors-total">{{ config.contributors.items() | length }}</span></span> {{ plural_ru(config.contributors.items() | length | int, ["участник", "участника", "участников"]) }}.{% endblock %}
```
Comment (dtikhonov) [inline src/templates/ru/2019/contributors.html:7]: > "участника", "участников", "участников"

File: src/templates/uk/2019/contributors.html [hunk 9]
```diff
@@ -2,9 +2,9 @@
...
+{% block description %}{{ config.contributors.items() | length }} {{ plural_ru(config.contributors.items() | length | int, ["неймовірна людина", "неймовірні людини", "неймовірних людей"]) }}, які долучилися до створення Web Almanac 2019 як аналітики, автори, дизайнери, розробники, рецензенти, керівники, редактори та перекладачі.{% endblock %}
 
-{% block filter_by_team %}Фільтрувати за роллю в команді: <span id="filtered-contributors">{{ self.contributors() }}</span><span id="contributors-total-text" class="hidden"> з <span id="contributors-total">{{ config.contributors.items() | length }}</span></span> контриб'юторів.{% endblock %}
+{% block filter_by_team %}Фільтрувати за роллю в команді: <span id="filtered-contributors">{{ self.contributors() }}</span><span id="contributors-total-text" class="hidden"> з <span id="contributors-total">{{ config.contributors.items() | length }}</span></span> {{ plural_ru(config.contributors.items() | length | int, ["контриб'ютор", "контриб'ютори", "контриб'юторів"]) }}.{% endblock %}
```
Comment (dtikhonov) [inline src/templates/uk/2019/contributors.html:7]: You missed a couple, @tymosh

File: src/templates/uk/2020/contributors.html [hunk 10]
```diff
@@ -2,9 +2,9 @@
...
+{% block description %}{{ config.contributors.items() | length }} {{ plural_ru(config.contributors.items() | length | int, ["неймовірна людина", "неймовірні людини", "неймовірних людей"]) }}, які долучилися до створення Web Almanac {{ year }} як аналітики, автори, дизайнери, розробники, рецензенти, керівники, редактори та перекладачі.{% endblock %}
 
-{% block filter_by_team %}Фільтрувати за роллю в команді: <span id="filtered-contributors">{{ self.contributors() }}</span><span id="contributors-total-text" class="hidden"> з <span id="contributors-total">{{ config.contributors.items() | length }}</span></span> контриб'юторів.{% endblock %}
+{% block filter_by_team %}Фільтрувати за роллю в команді: <span id="filtered-contributors">{{ self.contributors() }}</span><span id="contributors-total-text" class="hidden"> з <span id="contributors-total">{{ config.contributors.items() | length }}</span></span> {{ plural_ru(config.contributors.items() | length | int, ["контриб'ютор", "контриб'ютори", "контриб'юторів"]) }}.{% endblock %}
```
Comment (dtikhonov) [inline src/templates/uk/2020/contributors.html:7]: and this one.

**Review by dtikhonov — CHANGES_REQUESTED** [review]

**Comment by bazzadp** [comment]
> Thanks! Here's another thing: there are different word endings for "contributors" in constructions "124 contributors" and "5 of 124 contributors" in Ukrainian. Is there any way to do that?

Yes and have this working now. Take a look at the latest code. Can also view it here: https://20210204t153224-dot-webalmanac.uk.r.appspot.com

@dtikhonov I presume this is the same in Russian? Can you review the latest code?

File: src/templates/uk/2019/contributors.html (hunk 9 above)
Comment (bazzadp) [reply src/templates/uk/2019/contributors.html]: Resolving this as code has changed so let's pick up on latest code by doing another review (and ideally suggesting the changes using the suggest icon for my poor, non-Russian/non-Ukranian brain!).

File: src/templates/uk/2020/contributors.html (hunk 10 above)
Comment (bazzadp) [reply src/templates/uk/2020/contributors.html]: Resolving this as code has changed so let's pick up on latest code by doing another review (and ideally suggesting the changes using the suggest icon for my poor, non-Russian/non-Ukranian brain!).

File: src/templates/uk/2020/contributors.html [hunk 11]
```diff
@@ -2,9 +2,9 @@
...
+{% block description %}{{ config.contributors.items() | length }} {{ plural_ru(config.contributors.items() | length | int, ["неймовірна людина", "неймовірні людини", "неймовірних людей"]) }}, які долучилися до створення Web Almanac {{ year }} як аналітики, автори, дизайнери, розробники, рецензенти, керівники, редактори та перекладачі.{% endblock %}
 
-{% block filter_by_team %}Фільтрувати за роллю в команді: <span id="filtered-contributors">{{ self.contributors() }}</span><span id="contributors-total-text" class="hidden"> з <span id="contributors-total">{{ config.contributors.items() | length }}</span></span> контриб'юторів.{% endblock %}
+{% block filter_by_team %}Фільтрувати за роллю в команді: <span id="filtered-contributors">{{ self.contributors() }}</span><span id="contributors-total-text" class="hidden"> з <span id="contributors-total">{{ config.contributors.items() | length }}</span> {{ plural_ru(config.contributors.items() | length | int, ["контриб'ютор", "контриб'ютори", "контриб'юторів"]) }}</span><span id="contributors-total-text-non-filtered"> {{ plural_ru(config.contributors.items() | length | int, ["дописувач", "дописувачі", "дописувачів"]) }}</span>.{% endblock %}
```
Comment (tymosh) [inline src/templates/uk/2020/contributors.html:7]:
```suggestion
{% block filter_by_team %}Фільтрувати за роллю в команді: <span id="filtered-contributors">{{ self.contributors() }}</span><span id="contributors-total-text" class="hidden"> з <span id="contributors-total">{{ config.contributors.items() | length }}</span> {{ plural_ru(config.contributors.items() | length | int, ["контриб'ютора", "контриб'юторів", "контриб'юторів"]) }}</span><span id="contributors-total-text-non-filtered"> {{ plural_ru(config.contributors.items() | length | int, ["контриб'ютор", "контриб'ютори", "контриб'юторів"]) }}</span>.{% endblock %}
```

File: src/templates/uk/2019/contributors.html [hunk 12]
```diff
@@ -2,9 +2,9 @@
...
+{% block description %}{{ config.contributors.items() | length }} {{ plural_ru(config.contributors.items() | length | int, ["неймовірна людина", "неймовірні людини", "неймовірних людей"]) }}, які долучилися до створення Web Almanac 2019 як аналітики, автори, дизайнери, розробники, рецензенти, керівники, редактори та перекладачі.{% endblock %}
 
-{% block filter_by_team %}Фільтрувати за роллю в команді: <span id="filtered-contributors">{{ self.contributors() }}</span><span id="contributors-total-text" class="hidden"> з <span id="contributors-total">{{ config.contributors.items() | length }}</span></span> контриб'юторів.{% endblock %}
+{% block filter_by_team %}Фільтрувати за роллю в команді: <span id="filtered-contributors">{{ self.contributors() }}</span><span id="contributors-total-text" class="hidden"> з <span id="contributors-total">{{ config.contributors.items() | length }}</span> {{ plural_ru(config.contributors.items() | length | int, ["контриб'ютор", "контриб'ютори", "контриб'юторів"]) }}</span><span id="contributors-total-text-non-filtered">{{ plural_ru(config.contributors.items() | length | int, ["дописувач", "дописувачі", "дописувачів"]) }}</span>.{% endblock %}
```
Comment (tymosh) [inline src/templates/uk/2019/contributors.html:7]:
```suggestion
{% block filter_by_team %}Фільтрувати за роллю в команді: <span id="filtered-contributors">{{ self.contributors() }}</span><span id="contributors-total-text" class="hidden"> з <span id="contributors-total">{{ config.contributors.items() | length }}</span> {{ plural_ru(config.contributors.items() | length | int, ["контриб'ютора", "контриб'юторів", "контриб'юторів"]) }}</span><span id="contributors-total-text-non-filtered">{{ plural_ru(config.contributors.items() | length | int, ["контриб'ютор", "контриб'ютори", "контриб'юторів"]) }}</span>.{% endblock %}
```

**Review by tymosh — APPROVED** [review]
Looks good!

**Review by dtikhonov — APPROVED** [review]

**Pull request merged**